### PR TITLE
Link on test page to list of events related to run.

### PR DESCRIPTION
**Pull request by vdbergh**

### CODE REVIEW

**Comment by ppigazzini** [comment]
DEV updated with this PR as well.

**Comment by vdbergh** [comment]
It works but its a bit slow. I guess it is best to make an index on the `run_id` field.

**Comment by vdbergh** [comment]
I made an index on the run_id field (one should run `python create_indexes.py actions`).

**Comment by ppigazzini** [comment]
I experimented a bit trying to speedup the count of filtered runs using a covered query.
With mongosh the result is blazing fast, no documents inspected, no way to get that speedup with pymongo. I tried to count the length of the Cursors in python as well.

**Comment by vdbergh** [comment]
I think pymongo is just a thin layer above mongosh. Moreover I think one can use direct mongosh commands in pymongo as well (this is how the monitoring works IIRC). 

But in this case it is clear one needs an index on `run_id`, because otherwise we have to scan the whole event log. Of course the scanning uses an index but without an index on `run_id` we still have to fetch the records to filter them.

**Comment by ppigazzini** [comment]
Index added.

**Comment by vdbergh** [comment]
Now it is  very fast.

**Comment by ppigazzini** [comment]
I was talking about this count used for pagination.

https://github.com/glinscott/fishtest/blob/a9de4b1f94369328b7da9040fec974cb783c9dac/server/fishtest/rundb.py#L553-L557

Ideally we should simply do a covered query with a proper projection and ask for the count.
```python
 c = self.runs.find(q).skip(skip).limit(limit)
 count = self.runs.find(q, { "last_updated": -1, "_id": 0 }).count()
```

**Comment by vdbergh** [comment]
One can do "explain()" to check if indexes are used. It would be strange if mongosh would be faster than pymongo....

I think the problem is that the actiondb is so big that even an IXSCAN takes a lot of time.

I am busy now so can't look. But to speed up the "All" query one should create and index with a filter on the relevant events (all except dead_task, system_event and update_stats). In that way no IXSCAN is needed to count (I think an index knows its own size).

**Comment by ppigazzini** [comment]
I tested with explain. Only index used and no document scanned, as expected in a covered query.

**Comment by vdbergh** [comment]
> I tested with explain. Only index used and no document scanned, as expected in a covered query.

Yes but the index is still scanned (IXSCAN). I think that if the index is very large this still takes time. But I think that with a filter one can make it so that the count is the size of the index. I _think_ that an index knows its own size, without scanning.

**Comment by ppigazzini** [comment]
This count ran in millisecond in mongosh
```
fishtest_new> db.runs.find({"args.username": "Vizvezdenec"}, {"last_updated": -1, "_id": 0}).count()
19304
```

**Comment by ppigazzini** [comment]
```
fishtest_new> db.runs.find({"args.username": "Vizvezdenec"}, {"last_updated": -1, "_id": 0}).explain("executionStats")
{
  explainVersion: '1',
  queryPlanner: {
    namespace: 'fishtest_new.runs',
    indexFilterSet: false,
    parsedQuery: { 'args.username': { '$eq': 'Vizvezdenec' } },
    queryHash: '0D1F8E81',
    planCacheKey: '3C869B28',
    maxIndexedOrSolutionsReached: false,
    maxIndexedAndSolutionsReached: false,
    maxScansToExplodeReached: false,
    winningPlan: {
      stage: 'PROJECTION_COVERED',
      transformBy: { last_updated: -1, _id: 0 },
      inputStage: {
        stage: 'IXSCAN',
        keyPattern: { 'args.username': -1, last_updated: -1 },
        indexName: 'user_runs',
        isMultiKey: false,
        multiKeyPaths: { 'args.username': [], last_updated: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: {
          'args.username': [ '["Vizvezdenec", "Vizvezdenec"]' ],
          last_updated: [ '[MaxKey, MinKey]' ]
        }
      }
    },
    rejectedPlans: []
  },
  executionStats: {
    executionSuccess: true,
    nReturned: 19304,
    executionTimeMillis: 17,
    totalKeysExamined: 19304,
    totalDocsExamined: 0,
    executionStages: {
      stage: 'PROJECTION_COVERED',
      nReturned: 19304,
      executionTimeMillisEstimate: 1,
      works: 19305,
      advanced: 19304,
      needTime: 0,
      needYield: 0,
      saveState: 19,
      restoreState: 19,
      isEOF: 1,
      transformBy: { last_updated: -1, _id: 0 },
      inputStage: {
        stage: 'IXSCAN',
        nReturned: 19304,
        executionTimeMillisEstimate: 1,
        works: 19305,
        advanced: 19304,
        needTime: 0,
        needYield: 0,
        saveState: 19,
        restoreState: 19,
        isEOF: 1,
        keyPattern: { 'args.username': -1, last_updated: -1 },
        indexName: 'user_runs',
        isMultiKey: false,
        multiKeyPaths: { 'args.username': [], last_updated: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: {
          'args.username': [ '["Vizvezdenec", "Vizvezdenec"]' ],
          last_updated: [ '[MaxKey, MinKey]' ]
        },
        keysExamined: 19304,
        seeks: 1,
        dupsTested: 0,
        dupsDropped: 0
      }
    }
  },
  command: {
    find: 'runs',
    filter: { 'args.username': 'Vizvezdenec' },
    projection: { last_updated: -1, _id: 0 },
    '$db': 'fishtest_new'
  },
  serverInfo: {
    host: 'dfts-0',
    port: 27017,
    version: '6.0.3',
    gitVersion: 'f803681c3ae19817d31958965850193de067c516'
  },
  serverParameters: {
    internalQueryFacetBufferSizeBytes: 104857600,
    internalQueryFacetMaxOutputDocSizeBytes: 104857600,
    internalLookupStageIntermediateDocumentMaxSizeBytes: 104857600,
    internalDocumentSourceGroupMaxMemoryBytes: 104857600,
    internalQueryMaxBlockingSortMemoryUsageBytes: 104857600,
    internalQueryProhibitBlockingMergeOnMongoS: 0,
    internalQueryMaxAddToSetBytes: 104857600,
    internalDocumentSourceSetWindowFieldsMaxMemoryBytes: 104857600
  },
  ok: 1
}
```

**Comment by vdbergh** [comment]
Ah: I though tyou were talking about the actiondb. Getting to the last page takes an enormous amount of time. You want a COUNT_SCAN instead of an IXSCAN I think.

https://stackoverflow.com/questions/49554932/what-is-the-difference-between-count-scan-and-ixscan

**Comment by ppigazzini** [comment]
No, I want the count from the covered query as in mongosh, that simply inspect the value in the index.

**Comment by vdbergh** [comment]
Did you try talking directly to mongod? Like in the monitoring
```
monitoring = request.rundb.conn["admin"].command("getFreeMonitoringStatus")
```

**Comment by vdbergh** [comment]
I am curious. I think the issue with the actiondb is different (I think IXSCAN is too slow). But maybe not.

**Comment by ppigazzini** [comment]
> Did you try talking directly to mongod? Like in the monitoring
> 
> ```
> monitoring = request.rundb.conn["admin"].command("getFreeMonitoringStatus")
> ```

No, I will try.

**Comment by ppigazzini** [comment]
> I am curious. I think the issue with the actiondb is different (I think IXSCAN is too slow). But maybe not.

I think that the problem is the same, slow total count without a covered query.

**Comment by vdbergh** [comment]
The total count in the rundb is fast. So it seems different.

**Comment by ppigazzini** [comment]
> The total count in the rundb is fast. So it seems different.

Not when requiring a query, open this page:
https://tests.stockfishchess.org/tests/user/Vizvezdenec

**Comment by vdbergh** [comment]
Yes but that is a different issue I think (a bug in pymongo which makes it behave differently from mongosh). The query you are taking about _is_ covered (explain() only mentions an IXSCAN),

**Comment by ppigazzini** [comment]
For the total count we can always make a projection to have a covered query, the total count is coputed every time that we switch a page. The problem is that pymongo count the documents, skipping the index information.

**Comment by vdbergh** [comment]
This is the count for rundb. You see it uses a `COUNT_SCAN` which is faster than an `IXSCAN`.
```
fishtest_new> db.runs.explain().count({"finished": false})
{
  explainVersion: '1',
  queryPlanner: {
    namespace: 'fishtest_new.runs',
    indexFilterSet: false,
    parsedQuery: { finished: { '$eq': false } },
    queryHash: 'E8053644',
    planCacheKey: 'F15E782F',
    maxIndexedOrSolutionsReached: false,
    maxIndexedAndSolutionsReached: false,
    maxScansToExplodeReached: false,
    winningPlan: {
      stage: 'COUNT',
      inputStage: {
        stage: 'COUNT_SCAN',
        keyPattern: { finished: 1, last_updated: 1 },
        indexName: 'unfinished_runs',
        isMultiKey: false,
        multiKeyPaths: { finished: [], last_updated: [] },
        isUnique: false,
        isSparse: false,
        isPartial: true,
        indexVersion: 2,
        indexBounds: {
          startKey: { finished: false, last_updated: MinKey() },
          startKeyInclusive: true,
          endKey: { finished: false, last_updated: MaxKey() },
          endKeyInclusive: true
        }
      }
    },
    rejectedPlans: []
  },
  command: { count: 'runs', query: { finished: false }, '$db': 'fishtest_new' },
  serverInfo: {
    host: 'ip-172-31-50-113',
    port: 27017,
    version: '6.0.3',
    gitVersion: 'f803681c3ae19817d31958965850193de067c516'
  },
  serverParameters: {
    internalQueryFacetBufferSizeBytes: 104857600,
    internalQueryFacetMaxOutputDocSizeBytes: 104857600,
    internalLookupStageIntermediateDocumentMaxSizeBytes: 104857600,
    internalDocumentSourceGroupMaxMemoryBytes: 104857600,
    internalQueryMaxBlockingSortMemoryUsageBytes: 104857600,
    internalQueryProhibitBlockingMergeOnMongoS: 0,
    internalQueryMaxAddToSetBytes: 104857600,
    internalDocumentSourceSetWindowFieldsMaxMemoryBytes: 104857600
  },
  ok: 1
}
```

**Comment by vdbergh** [comment]
This is a corresponding query for the actiondb. It uses an IXSCAN since there is no index with the appropriate filter
```javascript
db.actions.explain().count({"action": {"$in": ["new_run", "stop_run"]}})
{
  explainVersion: '1',
  queryPlanner: {
    namespace: 'fishtest_new.actions',
    indexFilterSet: false,
    parsedQuery: { action: { '$in': [ 'new_run', 'stop_run' ] } },
    queryHash: '662765AF',
    planCacheKey: '30EE2D9B',
    maxIndexedOrSolutionsReached: false,
    maxIndexedAndSolutionsReached: false,
    maxScansToExplodeReached: false,
    winningPlan: {
      stage: 'COUNT',
      inputStage: {
        stage: 'IXSCAN',
        keyPattern: { action: 1, _id: -1 },
        indexName: 'action_1__id_-1',
        isMultiKey: false,
        multiKeyPaths: { action: [], _id: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: {
          action: [ '["new_run", "new_run"]', '["stop_run", "stop_run"]' ],
          _id: [ '[MaxKey, MinKey]' ]
        }
      }
    },
    rejectedPlans: []
  },
  command: {
    count: 'actions',
    query: { action: { '$in': [ 'new_run', 'stop_run' ] } },
    '$db': 'fishtest_new'
  },
  serverInfo: {
    host: 'ip-172-31-50-113',
    port: 27017,
    version: '6.0.3',
    gitVersion: 'f803681c3ae19817d31958965850193de067c516'
  },
  serverParameters: {
    internalQueryFacetBufferSizeBytes: 104857600,
    internalQueryFacetMaxOutputDocSizeBytes: 104857600,
    internalLookupStageIntermediateDocumentMaxSizeBytes: 104857600,
    internalDocumentSourceGroupMaxMemoryBytes: 104857600,
    internalQueryMaxBlockingSortMemoryUsageBytes: 104857600,
    internalQueryProhibitBlockingMergeOnMongoS: 0,
    internalQueryMaxAddToSetBytes: 104857600,
    internalDocumentSourceSetWindowFieldsMaxMemoryBytes: 104857600
  },
  ok: 1
}
```

**Comment by vdbergh** [comment]
I think there are really two different issues here.

**Comment by ppigazzini** [comment]
Add "executionStats" to get more info:

```bash
db.runs.explain("executionStats").count({"args.username": "Vizvezdenec"})
{
  explainVersion: '1',
  queryPlanner: {
    namespace: 'fishtest_new.runs',
    indexFilterSet: false,
    parsedQuery: { 'args.username': { '$eq': 'Vizvezdenec' } },
    queryHash: 'A8602F55',
    planCacheKey: '4D8E47DE',
    maxIndexedOrSolutionsReached: false,
    maxIndexedAndSolutionsReached: false,
    maxScansToExplodeReached: false,
    winningPlan: {
      stage: 'COUNT',
      inputStage: {
        stage: 'COUNT_SCAN',
        keyPattern: { 'args.username': -1, last_updated: -1 },
        indexName: 'user_runs',
        isMultiKey: false,
        multiKeyPaths: { 'args.username': [], last_updated: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        indexBounds: {
          startKey: { 'args.username': 'Vizvezdenec', last_updated: MaxKey() },
          startKeyInclusive: true,
          endKey: { 'args.username': 'Vizvezdenec', last_updated: MinKey() },
          endKeyInclusive: true
        }
      }
    },
    rejectedPlans: []
  },
  executionStats: {
    executionSuccess: true,
    nReturned: 0,
    executionTimeMillis: 8,
    totalKeysExamined: 19305,
    totalDocsExamined: 0,
    executionStages: {
      stage: 'COUNT',
      nReturned: 0,
      executionTimeMillisEstimate: 0,
      works: 19305,
      advanced: 0,
      needTime: 19304,
      needYield: 0,
      saveState: 19,
      restoreState: 19,
      isEOF: 1,
      nCounted: 19304,
      nSkipped: 0,
      inputStage: {
        stage: 'COUNT_SCAN',
        nReturned: 19304,
        executionTimeMillisEstimate: 0,
        works: 19305,
        advanced: 19304,
        needTime: 0,
        needYield: 0,
        saveState: 19,
        restoreState: 19,
        isEOF: 1,
        keysExamined: 19305,
        keyPattern: { 'args.username': -1, last_updated: -1 },
        indexName: 'user_runs',
        isMultiKey: false,
        multiKeyPaths: { 'args.username': [], last_updated: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        indexBounds: {
          startKey: { 'args.username': 'Vizvezdenec', last_updated: MaxKey() },
          startKeyInclusive: true,
          endKey: { 'args.username': 'Vizvezdenec', last_updated: MinKey() },
          endKeyInclusive: true
        }
      }
    }
  },
  command: {
    count: 'runs',
    query: { 'args.username': 'Vizvezdenec' },
    '$db': 'fishtest_new'
  },
  serverInfo: {
    host: 'dfts-0',
    port: 27017,
    version: '6.0.3',
    gitVersion: 'f803681c3ae19817d31958965850193de067c516'
  },
  serverParameters: {
    internalQueryFacetBufferSizeBytes: 104857600,
    internalQueryFacetMaxOutputDocSizeBytes: 104857600,
    internalLookupStageIntermediateDocumentMaxSizeBytes: 104857600,
    internalDocumentSourceGroupMaxMemoryBytes: 104857600,
    internalQueryMaxBlockingSortMemoryUsageBytes: 104857600,
    internalQueryProhibitBlockingMergeOnMongoS: 0,
    internalQueryMaxAddToSetBytes: 104857600,
    internalDocumentSourceSetWindowFieldsMaxMemoryBytes: 104857600
  },
  ok: 1
}
```

**Comment by vdbergh** [comment]
I did not add ExectutionStats since I was only interested in the query plan!

**Comment by ppigazzini** [comment]
Yes, but you should check that the query plan is working as intended :)

**Comment by ppigazzini** [comment]
BTW the in mongosh the count of Viz's runs is blazing fast as well without a covered query...

**Comment by ppigazzini** [comment]
I missed the the real query involves "finished", making the count slow in mongosh as well:

`db.runs.explain("executionStats").find({"finished": true, "args.username": "Vizvezdenec"}, {"last_updated": -1, "_id":0})`

**Comment by vdbergh** [comment]
> I missed that the real query involves "finished", making the count slow in mongosh as well because mongodb must examine the documents.
> 
> ```python
> db.runs.explain("executionStats").find({"finished": true, "args.username": "Vizvezdenec"}, {"last_updated": -1, "_id":0})
> 
> db.runs.explain("executionStats").count({"finished": true, "args.username": "Vizvezdenec"})
> 
>   executionStats: {
>     executionSuccess: true,
>     nReturned: 0,
>     executionTimeMillis: 5059,
>     totalKeysExamined: 19304,
>     totalDocsExamined: 19304,
>     executionStages: {
>       stage: 'COUNT',
>       nReturned: 0,
>       executionTimeMillisEstimate: 4895,
>  
> ```

It seems we are missing an index on username with filter 'finished':true ...
```
fishtest_new> indexes runs 
_id_                          {'_id': 'asc'}                
unfinished_runs               {'finished': 'asc', 'last_updated': 'asc'}         {'finished': False}
finished_runs                 {'finished': 'asc', 'last_updated': 'desc'}         {'finished': True}
finished_green_runs           {'finished': 'asc', 'is_green': 'desc', 'last_updated': 'desc'}         {'finished': True, 'is_green': True}
finished_yellow_runs          {'finished': 'asc', 'is_yellow': 'desc', 'last_updated': 'desc'}         {'finished': True, 'is_yellow': True}
finished_ltc_runs             {'finished': 'asc', 'last_updated': 'desc', 'tc_base': 'desc'}         {'finished': True, 'tc_base': {'$gte': 40}}
user_runs                     {'args.username': 'desc', 'last_updated': 'desc'}
```
(output of MiniMongoShel)l.

**Comment by vdbergh** [comment]
An index on username+finished (possibly with a filter as well) should solve the problem!

**Comment by ppigazzini** [comment]
> An index on username+finished (possibly with a filter as well) should solve the problem!

Please write the PR, I'm a noob in mongodb :)

**Comment by vdbergh** [comment]
Ok I'll try...

**Comment by vdbergh** [comment]
So it seems that 
```javascript
db.actions.explain().count({"action": {"$in":["stop_run"]}})
```
does a COUNT_SCAN and
```javascript
db.actions.explain().count({"action": {"$in":["stop_run","new_run"]}})
```
does an IXSCAN. I wonder if mongodb decides that one IXSCAN is cheaper than two COUNT_SCANS?

File: server/fishtest/templates/tests_view.mak [hunk 1]
```diff
@@ -122,6 +122,10 @@
                 </tr>
             % endif
         % endfor
+	<tr>
+	  <td>events</td>
+	  <td><a href="/actions?run_id=${str(run['_id'])}">/actions?run_id=${run['_id']}</a></td>
+	<tr>
```
Comment (ppigazzini) [inline server/fishtest/templates/tests_view.mak:128]: typo in closing tag `</tr>`

Comment (vdbergh) [reply server/fishtest/templates/tests_view.mak]: Fixed!

Comment (dav1312) [reply server/fishtest/templates/tests_view.mak]: also tabs

Comment (vdbergh) [reply server/fishtest/templates/tests_view.mak]: Also fixed.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated, thank you @vdbergh :)

**Comment by ppigazzini** [comment]
I forgot to update the index, doing now on PROD...

**Comment by vdbergh** [comment]
There is still a bug if a run spans more than one page in the event log. The page links and the links to individual items do not take into account the new run_id parameters. I will fix it after new year.

**Pull request merged**